design/one-pager-function-go-templating.md

```diff
@@ -0,0 +1,183 @@
+template engine. This will allow users to access resource fields in the 
+template files like:
+
+```yaml
+{{- range $i := until ( .observed.composite.resource.spec.count | int ) }}
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: User
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: test-user-{{ $i }}
+  name: test-user-{{ $i }}
+  labels:
+    testing.upbound.io/example-name: test-user-{{ $i }}
+spec:
+  forProvider: {}
+
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: AccessKey
+metadata:
+  name: sample-access-key-{{ $i }}
+spec:
+  forProvider:
+    userSelector:
+      matchLabels:
+        testing.upbound.io/example-name: test-user-{{ $i }}
+  writeConnectionSecretToRef:
+    name: sample-access-key-secret-{{ $i }}
+    namespace: crossplane-system
+{{- end }}
+
+```
+
+To compare and prepare the desired composed resources map with consistent keys,
+`function-go-templating` requires a special annotation 
+`gotemplating.fn.crossplane.io/composition-resource-name` to be assigned to the
+input resources. This annotation will match the resource name in the template
+file with the resource name in the desired resources.
+
+## Configuring Composite Connection Details
+
+A new meta resource, `CompositeConnectionDetails` with API version 
+`meta.gotemplating.fn.crossplane.io/v1alpha1` will be used to set connection 
+details of the composite resource. This resource will contain a `data` field,
+which expects a map with string keys and base64 encoded string values. This 
+meta resource is not a real Crossplane resource, and it will not be applied to
+the cluster.
+
+```yaml
+apiVersion: meta.gotemplating.fn.crossplane.io/v1alpha1
+kind: CompositeConnectionDetails
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: connection-details
+  name: connection-details
+{{ if eq $.observed.resources nil }}
+data: {}
+{{ else }}
+data:
+  username: {{ ( index $.observed.resources "sample-access-key-0" ).connectionDetails.username }}
+  password: {{ ( index $.observed.resources "sample-access-key-0" ).connectionDetails.password }}
+  url: {{ "http://www.example.com" | b64enc }}
+{{ end }}
+
+```
+
+## Configuring Resource Status
+
+In addition to that, a new custom annotation, 
+`gotemplating.fn.crossplane.io/ready` will be introduced to set the ready 
+status of the resource, which can be assigned to `Unspecified`, `True` or 
+`False`. This annotation will be removed after the resource is rendered and not
+applied to the cluster.
+
+```yaml
+apiVersion: kubernetes.crossplane.io/v1alpha1
+kind: ProviderConfig
+metadata:
+  name: composed-providerconfig
+  annotations:
+    gotemplating.fn.crossplane.io/ready: True
+spec:
+  credentials:
+    source: Secret
+    secretRef:
+      namespace: crossplane-system
+      name: cluster-config
+      key: kubeconfig
+```
+
+## Helper Functions
+
+A set of helper functions like `getComposedResource` or `getEnvVar` will be 
```

Comment:
I think `getEnvVar` will be confusing - the suggestion to use `getEnvConfigVar` or something similar is more descriptive.

Reply:
Discussed with @phisco offline. Apparently, there may be more key/value pairs coming apart from `environmentConfig` resources with `"apiextensions.crossplane.io/environment"` annotation. Therefore using `getCompositionEnvVar` makes more sense here. However, we should definitely document `Composition Environment` and try to prevent the confusion.

Reply:
Don't forget it's an alpha feature. Clarifying the terms I think will play a key role in the promotion to beta of the functionality.